dali/pipeline/operators/transpose/transpose.cu

```diff
@@ -30,20 +31,57 @@ namespace dali {
   }                                                            \
 } while (0)
 
-using IntArr = std::unique_ptr<int[]>;
+namespace detail {
 
-std::pair<IntArr, IntArr>
-RowToColumnMajor(const int *dims, const int *perm, int rank) {
-  IntArr new_dims(new int[rank]);
-  IntArr new_perm(new int[rank]);
+void RowToColumnMajor(int *dims,
+                      int *perm,
+                      size_t len) {
+  std::reverse(dims, dims + len);
+  std::reverse(perm, perm + len);
+  for (size_t i = 0; i < len; ++i)
+    perm[i] = len - 1 - perm[i];
+}
 
-  for (int i = 0; i < rank; ++i) {
-    new_dims[i] = dims[rank - 1 - i];
-    new_perm[i] = rank - 1 - perm[rank - 1 - i];
+// enough to represent batches of 4D
+using VecInt = SmallVector<int, 5>;
+
+void PrepareArguments(VecInt &shape,
+                      VecInt &perm) {
+  DALI_ENFORCE(shape.size() == perm.size());
+
+  VecInt idx;
+  idx.resize(shape.size());
+  std::iota(idx.begin(), idx.end(), 0);
+
+  // cuTT does not handle dimensions with size 1 so we remove them
+  // (H, W, 1) is equivalent to (H, W)
+  auto it_shape = shape.begin();
+  auto it_perm = perm.begin();
+  auto it_idx = idx.begin();
+  while (it_shape != shape.end()) {
+    if (*it_shape == 1) {
+      it_shape = shape.erase(it_shape);
+      it_perm = perm.erase(it_perm);
+      it_idx = idx.erase(it_idx);
+    } else {
+      ++it_shape;
+      ++it_perm;
+      ++it_idx;
+    }
   }
-  return {std::move(new_dims), std::move(new_perm)};
+
+  std::unordered_map<int, int> idx_map;
```

Comment:
Possible improvement: this could be just a vector - although you'd need to shift args by 1.

Reply:
done